### DISCUSSION TITLE
Are you ready to begin embedding smart blocks - each time I open Obsidian

### DISCUSSION THREAD

**Discussion by nomadphase** [brianpetro/obsidian-smart-connections#469] [General]
Hi Brian,

![image](https://github.com/brianpetro/obsidian-smart-connections/assets/72380765/858b3ded-25c0-4e48-a360-d944a2a91cde)

I am being prompted to begin embedding smart blocks each time I open Obsidian. My understanding was that this would be done once when upgrading to V2 (and maybe also for a couple of additional updates since). Is it typical that it should be happening each time I open Obsidian (typically once per day).

Cheers,

**Comment by arminta7** [comment]
I am actually getting this every time I switch notes and it locks up Obsidian for about 5 secs or so.

**Reply by brianpetro** [reply]
@arminta7 I noticed this issue, too. 

I started looking into it earlier today, but I still need to fix it.

I'm going to bump the priority and I'll let you know when it's fixed.

🌴

**Reply by brianpetro** [reply]
@arminta7 I found an error in the logic that I am pretty sure was causing this issue.

Instead of just embedding the current note if it wasn't already embedded, embedding all un-embedded entities was triggered.

This should be fixed in `v2.0.115`.

Let me know if it keeps happening and I'll give it another look.

Thanks for your help,
🌴

**Comment by brianpetro** [comment]
@nomadphase that is not supposed to be happening.

I'm wondering if the embedding file isn't saving correctly. 

Do you think that 12243 blocks represents all of the blocks in your vault? 

Which model are you using?

🌴

**Reply by nomadphase** [reply]
Thanks Brian,

My main vault is 14.4 million words spread across 2890 notes. However, due to my exclusion criteria policies, the embeddings include 1390 files. 

I am using OpenAI Text-3 Large for notes and block level embedding.

![image](https://github.com/brianpetro/obsidian-smart-connections/assets/72380765/49bedb1d-7200-4b4c-aeb9-7604c344d090)

When I was using v1, my embeddings file was about 500MB based on this vault, give or take some small changes.

**Reply by nomadphase** [reply]
so, it looks like the embeddings are there and I haven't made any massive changes to the vault, like a folder re-structure. It's really a matter of booting up my computer, switching on obsidian and being prompted to reinitiate the embeddings process.

**Reply by brianpetro** [reply]
@nomadphase can you try this version of the Large model with few dimensions?

![Screenshot 2024-02-18 at 10 11 08 PM](https://github.com/brianpetro/obsidian-smart-connections/assets/1886014/04c5a373-64e1-4ac8-a495-737b0a768360)

It shouldn't be behaving the way you're saying, so my guess is that the embeddings aren't fully saving. Switching to the model with reduced dimensions (the 256 dimensions still performs better than the previous Ada model according to OpenAI) should reduce the file size enough to get things working immediately.

🌴

**Reply by nomadphase** [reply]
that is interesting. however, when I ask a question of the smart connections chat, it is providing me with detailed responses, so it is able to access an embeddings file (I don't know which one). 

i guess i'm weighing up the trade-off between the model performance and the glitch prompt to re-embed the vault?

**Reply by brianpetro** [reply]
@nomadphase Did you try changing the models?

The chat is still working with the higher dimension model because all of the notes are being re-embedded and held in memory, and the embeddings in memory are used during the lookup process. But these are probably failing to be saved to the disk, which if why it's re-embedding frequently.

If you're wondering whether the reduced dimensions will reduce performance, my guess is that the difference in performance will be insignificant. So, I highly recommend at least trying it out.

🌴

**Reply by nomadphase** [reply]
Thanks Brian,

I've begun re-embedding with a different model - the 1536 dimension model

![image](https://github.com/brianpetro/obsidian-smart-connections/assets/72380765/aefbd108-e8d3-42fa-b845-512f1d20961d)

My setup is beefy: i914900k, 128GB Ram, Nvidia 4090 - so I should be able to run a very powerful local model...

**Reply by brianpetro** [reply]
@nomadphase, impressive setup!

Unfortunately, an internal feature in JavaScript (`JSON.stringify`) seems to break at ~500MB, regardless of the machine specs. And that's the issue I believe we're running into.

🌴

**Reply by nomadphase** [reply]
aha - so hopefully the 1536 dimension model will come in below that constraint. 

Hopefully there will be a way to overcome that constraint in the longer term, as I expect people's vaults and RAG embeddings will grow over time, even if I am currently a 'power user'.

Thank you for your support!

**Reply by brianpetro** [reply]
@nomadphase it's definitely on the roadmap and it should be done sooner than later.

There are a few different ways to go about it, and I'm still deciding which will make the most sense.

I know it's a frustrating issue, so thanks for bearing with me until it gets fixed!

🌴

**Reply by nomadphase** [reply]
Thanks Brian,

I can see you've been active on this issue, I saw your posts on the OpenAI forum from back in July!

I'm just keen to have the best information retrieval possible, whether it is local or hosted. It's important that it is reliable and that it outperforms other knowledge retrieval tech/tools. 

All in all, Obsidian, Dataview, Tasks, Omnisearch, Properties, Smart Connections - it all gives me an amazing set of tools to work with a massive amount of information - it's game changing

**Reply by nomadphase** [reply]
Looks like the 1536 dim model is the go:

![image](https://github.com/brianpetro/obsidian-smart-connections/assets/72380765/f6bfbef4-e052-4e21-98b1-904142a2224c)

I shut down and restarted Obsidian and was prompted to embed 3 smart blocks! win

**Comment by SoulBits-Vibe** [comment]
I have been getting the embedding message very often recently and looked at my settings and realized I don't understand any of this yet. 
Thought it might be helpful to ask if I should change these settings?

![image](https://github.com/brianpetro/obsidian-smart-connections/assets/61413087/ade0dfaf-4294-4e7d-bf10-6d997f79d2c1)
![image](https://github.com/brianpetro/obsidian-smart-connections/assets/61413087/a3796b59-66ab-4346-a3c7-c42d91185deb)

**Reply by brianpetro** [reply]
Hi @destiny911 

No, that looks good.

The only thing you might try is setting blocks to "none" and notes to the BGE-micro model, or maybe try an OpenAI model.

But I think the issue is primarily two things that I'm in the process of addressing:

1) some logic errors in how "detecting changes notes" notes is implemented.
2) excess notifications in general

Both of these issues will be addressed in the next update. I already made some tweaks to the logic, and added the ability to "mute" notifications. 

I just need to do a little more testing before I release the update.

🌴

**Reply by Derducken** [reply]
Hello @brianpetro , any news on a fix for the issue? My Ryzen 5900x with an RTX 3070 and a local LLM goes at around 700 tokens per second, and in a vault with around 3000 long-ish notes, the process seems to run whenever I launch Obsidian, locking it up for much, much more than "seconds". As I'm writing this, it's been over ten minutes I'm waiting for it to re-process the same notes, and it's only went through less than 1/3rd of their total.

**Reply by brianpetro** [reply]
Hi @Derducken 

The embedding speed will depend a lot on which model you are using. That is slow, but based on your hardware, I'm guessing you selected a larger model other than the default BGE-micro, which typically makes it through my ~3000 notes in < 10 minutes. 

The other thing to note is that the local embedding models are not GPU-optimized. This is a factor of using HuggingFace transformers.js, which currently doesn't support GPUs, but this should be available in the future and I'm watching the progress closely.

Lastly, the re-embedding issue can be followed here https://github.com/brianpetro/obsidian-smart-connections/issues/496#issuecomment-2033801234 

🌴

**Reply by Derducken** [reply]
Hello @brianpetro , thanks for that, but unfortunately I'm not using the supporter version. So, I'm stuck with having to recreate everything from scratch multiple times per day. Using a smaller model does make the process go faster, but also dramatically drops the "quality" of the suggestions (although that's subjective).

I guess I'll have to wait until your fixes and additions (like the new "per note" setting) reaches the "main" publicly available version.
